contexts/PaymentContext/src/Domain/ChecksumGenerator.php

```diff
@@ -0,0 +1,43 @@
+<?php
+
+declare( strict_types = 1 );
+
+namespace WMDE\Fundraising\Frontend\PaymentContext\Domain;
+
+/**
+ * @licence GNU GPL v2+
+ */
+class ChecksumGenerator {
+
+	private $checksumCharacters;
+
+	/**
+	 * @param array $checksumCharacters Characters that can be used for the checksum
+	 */
+	public function __construct( array $checksumCharacters ) {
+		if ( count( $checksumCharacters ) < 2 ) {
+			throw new \InvalidArgumentException(
+				'Need at least two characters to create meaningful checksum'
+			);
+		}
+
+		$this->checksumCharacters = $checksumCharacters;
+	}
+
+	/**
+	 * @param string $string The string to create a checksum for
+	 *
+	 * @return string The checksum as a single character present in the constructors array argument
+	 */
+	public function createChecksum( string $string ): string {
+		$checksum = array_sum(
+			array_map(
+				'ord', // Returns the ASCII value of the character
+				str_split( md5( $string ) )
+			)
```

Comment:
This will produce a very limited range of low numbers between 1563 ( `ord('0') * 32` ) and 2496 ( `ord('F') * 32` ), equating a range between 0-960, [that's about 10 bits](https://stackoverflow.com/a/23400205/130121). The more bits, the more accurate the checksum. How about 

    base_convert( substr( md5( $string ), 0, 16 ), 16, 10 );

the `substr` will "convert" the 32-bit hash into a 16-bit hash, `base_convert` converts the hex string into an integer between 0 and 18446744073709552046

Reply:
Does that matter if we `% 16` it later anyway? I checked the distribution created for the checksum and it was pretty even.

Reply:
It probably does not matter. However, as this algorithm has to be implemented elsewhere in some other language and documented, I'd like it to be as "mathematical" as possible and not to rely on string operations and ASCII conversion. `base_convert(substr( md5($string), 0, 16))` could also be expressed as "shift the bits of the MD5 hash integer value to the right by 64 places".

BTW, I confused bits and characters in my comment above. The truncation of 16 characters converts the 128-bit hash (expressed as 32 hex characters) into a 64-bit hash. Which worked on my machine because there PHP integers have 64 bits, but on some machines they only have 32, which is why I suppose we bit-shift the md5 hash to 32 bits:

    base_convert( substr( md5( $string ), 0, 8 ), 16, 10 );